<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Nota" data-toc-modified-id="Nota-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Nota</a></span></li></ul></div>

# Déjame escuchar la música

# Tabla de contenidos <a id='back'></a>

* [Introducción](#intro)
* [Etapa 1. Descripción de los datos](#data_review)
    * [Conclusiones](#data_review_conclusions)
* [Etapa 2. Preprocesamiento de datos](#data_preprocessing)
    * [2.1 Estilo del encabezado](#header_style)
    * [2.2 Valores ausentes](#missing_values)
    * [2.3 Duplicados](#duplicates)
    * [2.4 Conclusiones](#data_preprocessing_conclusions)
* [Etapa 3. Prueba de hipótesis](#hypothesis)
    * [3.1 Hipótesis 1: actividad de los usuarios y las usuarias en las dos ciudades](#activity)
* [Conclusiones](#end)

# Introducción <a id="intro"></a>
Como analista de datos, tu trabajo consiste en analizar datos para extraer información valiosa y tomar decisiones basadas en ellos. Esto implica diferentes etapas, como la descripción general de los datos, el preprocesamiento y la prueba de hipótesis.

Siempre que investigamos, necesitamos formular hipótesis que después podamos probar. A veces aceptamos estas hipótesis; otras veces, las rechazamos. Para tomar las decisiones correctas, una empresa debe ser capaz de entender si está haciendo las suposiciones correctas.

En este proyecto, compararás las preferencias musicales de las ciudades de Springfield y Shelbyville. Estudiarás datos reales de transmisión de música online para probar la hipótesis a continuación y comparar el comportamiento de los usuarios y las usuarias de estas dos ciudades.

Objetivo:
Prueba la hipótesis:
1. La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad.

Etapas
Los datos del comportamiento del usuario se almacenan en el archivo `/datasets/music_project_en.csv`. No hay ninguna información sobre la calidad de los datos, así que necesitarás examinarlos antes de probar la hipótesis.

Primero, evaluarás la calidad de los datos y verás si los problemas son significativos. Entonces, durante el preprocesamiento de datos, tomarás en cuenta los problemas más críticos.

Tu proyecto consistirá en tres etapas:
 1. Descripción de los datos.
 2. Preprocesamiento de datos.
 3. Prueba de hipótesis.








[Volver a Contenidos](#back)

# Etapa 1. Descripción de los datos<a id='data_review'></a>

Abre los datos y examínalos.

Necesitarás `pandas`, así que impórtalo.

In [2]:
# Importar pandas
import pandas as pd

Lee el archivo `music_project_en.csv` de la carpeta `/datasets/` y guárdalo en la variable `df`:

In [12]:
# Leer el archivo CSV desde la carpeta /datasets/
file_path = '/datasets/music_project_en.csv'
df = pd.read_csv(file_path)

Muestra las 10 primeras filas de la tabla:

In [6]:
# Obtener las 10 primeras filas de la tabla df
print(df.head(10))

     userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5

Obtén la información general sobre la tabla con un comando. Conoces el método que muestra la información general que necesitamos.

In [8]:
# Obtener la información general sobre nuestros datos
df_info = df.info()

# Imprimir la información general
print(df_info)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB
None


Estas son nuestras observaciones sobre la tabla. Contiene siete columnas. Almacenan los mismos tipos de datos: `object`.

Según la documentación:
- `' userID'`: identificador del usuario o la usuaria;
- `'Track'`: título de la canción;
- `'artist'`: nombre del artista;
- `'genre'`: género de la pista;
- `'City'`: ciudad del usuario o la usuaria;
- `'time'`: la hora exacta en la que se reprodujo la canción;
- `'Day'`: día de la semana.

Podemos ver tres problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros en minúsculas.
2. Hay espacios en algunos encabezados.
3. Inconsistencia en el uso de los espacios y el formato general en los nombres de las columnas.



# Conclusiones<a id='data_review_conclusions'></a>

Tipo de Datos en las Filas y Entendimiento de las Columnas:
Al revisar los datos, me di cuenta de que cada fila representa la reproducción de una canción por un usuario en una ciudad específica y en un día particular de la semana. Las columnas nos brindan información detallada: userID es el identificador del usuario, track es el título de la canción, artist es el nombre del artista, genre es el género musical, city es la ciudad del usuario, time es la hora exacta en que se reprodujo la canción, y day es el día de la semana. Todos estos datos están en formato de texto, lo cual tiene sentido dado el tipo de información.

Suficiencia de Datos para la Hipótesis:
Para asegurarme de que tenemos suficientes datos para probar nuestra hipótesis, revisé la cantidad total de registros y cómo se distribuyen entre las diferentes ciudades y días de la semana. Encontré que el conjunto de datos es bastante amplio y que hay una buena representación de usuarios de cada ciudad y de cada día de la semana. Esto me da la confianza de que tenemos suficiente información para proceder con nuestra prueba de hipótesis.

Problemas en los Datos:
Durante mi revisión, identifiqué algunos problemas en los datos. Por ejemplo, hay valores ausentes en algunas columnas, aunque no son muchos. También encontré filas duplicadas que podrían afectar nuestros resultados y que deberíamos eliminar. Además, noté que los nombres de las columnas no son consistentes: algunos tienen espacios en blanco y hay una mezcla de mayúsculas y minúsculas. Planeo corregir estos problemas en el preprocesamiento de datos para asegurarme de que todo esté en orden antes de seguir adelante con el análisis.

Resumen:
En resumen, aunque los datos son en su mayoría adecuados para nuestro análisis, necesitan algunos ajustes. Creo que tenemos una cantidad y diversidad suficientes para probar nuestra hipótesis sobre las diferencias en la actividad de los usuarios según la ciudad y el día de la semana.

[Volver a Contenidos](#back)

# Etapa 2. Preprocesamiento de datos <a id='data_preprocessing'></a>

El objetivo aquí es preparar los datos para que sean analizados.
El primer paso es resolver cualquier problema con los encabezados. Luego podemos avanzar a los valores ausentes y duplicados. Empecemos.

Corrige el formato en los encabezados de la tabla.


# 2.1 Estilo del encabezado <a id='header_style'></a> 

Muestra los encabezados de la tabla (los nombres de las columnas):

In [10]:
# Muestra los nombres de las columnas
# Mostrar los nombres de las columnas originales
print(df.columns)


Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


Cambia los encabezados de la tabla de acuerdo con las reglas del buen estilo:
* Todos los caracteres deben ser minúsculas.
* Elimina los espacios.
* Si el nombre tiene varias palabras, utiliza snake_case.

In [16]:
# Utilizar un bucle for para iterar sobre los nombres de las columnas y poner todos los caracteres en minúsculas
new_columns = []
for col in df.columns:
    new_columns.append(col.lower())

# Asignar los nuevos nombres de las columnas al dataframe
df.columns = new_columns

# Mostrar los nombres de las columnas después de la normalización
print("Nombres normalizados de las columnas:")
print(df.columns)


Nombres normalizados de las columnas:
Index(['userid', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


Ahora, utilizando el mismo método, elimina los espacios al principio y al final de los nombres de las columnas e imprime los nombres de las columnas nuevamente:

In [15]:
# Utilizar un bucle for para iterar sobre los nombres de las columnas,
# poner todos los caracteres en minúsculas y eliminar los espacios al principio y al final
new_columns = []
for col in df.columns:
    new_columns.append(col.strip().lower())

# Asignar los nuevos nombres de las columnas al dataframe
df.columns = new_columns

# Mostrar los nombres de las columnas después de la normalización
print("Nombres normalizados de las columnas:")
print(df.columns)


Nombres normalizados de las columnas:
Index(['userid', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


Necesitamos aplicar la regla de snake_case a la columna `userid`. Debe ser `user_id`. Cambia el nombre de esta columna y muestra los nombres de todas las columnas cuando hayas terminado.

In [18]:
# Asignar los nuevos nombres de las columnas al dataframe
df.columns = new_columns

# Cambiar el nombre de la columna 'userid' a 'user_id'
df.rename(columns={'userid': 'user_id'}, inplace=True)

# Mostrar los nombres de las columnas después de la normalización
print("Nombres normalizados de las columnas:")
print(df.columns)


Nombres normalizados de las columnas:
Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


Comprueba el resultado. Muestra los encabezados una vez más:

In [55]:
# Comprobar el resultado: la lista de encabezados

# Mostrar los nombres de las columnas después de la normalización
print("Nombres normalizados de las columnas:")
print(df.columns)


Nombres normalizados de las columnas:
Index(['userid', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Muy buen trabajo! Ajustaste todos los elementos</div>

[Volver a Contenidos](#back)

# 2.2 Valores ausentes <a id='missing_values'></a>
 Primero, encuentra el número de valores ausentes en la tabla. Debes utilizar dos métodos en una secuencia para obtener el número de valores ausentes.

In [22]:
# Calcular el número de valores ausentes

# Calcular el número de valores ausentes en cada columna
missing_values = df.isnull().sum()

# Mostrar el número de valores ausentes
print("Número de valores ausentes en cada columna:")
print(missing_values)


Número de valores ausentes en cada columna:
  userID       0
Track       1343
artist      7567
genre       1198
  City         0
time           0
Day            0
dtype: int64


No todos los valores ausentes afectan a la investigación. Por ejemplo, los valores ausentes en `track` y `artist` no son cruciales. Simplemente puedes reemplazarlos con valores predeterminados como el string `'unknown'` (desconocido).

Pero los valores ausentes en `'genre'` pueden afectar la comparación entre las preferencias musicales de Springfield y Shelbyville. En la vida real, sería útil saber las razones por las cuales hay datos ausentes e intentar recuperarlos. Pero no tenemos esa oportunidad en este proyecto. Así que tendrás que:
* rellenar estos valores ausentes con un valor predeterminado;
* evaluar cuánto podrían afectar los valores ausentes a tus cómputos;

Reemplazar los valores ausentes en las columnas `'track'`, `'artist'` y `'genre'` con el string `'unknown'`. Como mostramos anteriormente en las lecciones, la mejor forma de hacerlo es crear una lista que almacene los nombres de las columnas donde se necesita el reemplazo. Luego, utiliza esta lista e itera sobre las columnas donde se necesita el reemplazo haciendo el propio reemplazo.

In [25]:
# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'

# Normalizar los nombres de las columnas
df.columns = [col.strip().lower() for col in df.columns]

# Mostrar los nombres de las columnas actuales
print("Nombres normalizados de las columnas:")
print(df.columns)

# Lista de columnas donde se necesita el reemplazo
# Ajustar esta lista según los nombres correctos de las columnas
columns_to_replace = ['track', 'artist', 'genre']

# Iterar sobre las columnas y reemplazar los valores ausentes con 'unknown'
for col in columns_to_replace:
    df[col].fillna('unknown', inplace=True)

# Verificar que los valores ausentes han sido reemplazados
missing_values_after_replacement = df.isnull().sum()

# Mostrar el número de valores ausentes después del reemplazo
print("Número de valores ausentes en cada columna después del reemplazo:")
print(missing_values_after_replacement)


Nombres normalizados de las columnas:
Index(['userid', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')
Número de valores ausentes en cada columna después del reemplazo:
userid    0
track     0
artist    0
genre     0
city      0
time      0
day       0
dtype: int64


Ahora comprueba el resultado para asegurarte de que después del reemplazo no haya valores ausentes en el conjunto de datos. Para hacer esto, cuenta los valores ausentes nuevamente.

In [27]:
# Contar valores ausentes


# Verificar que los valores ausentes han sido reemplazados
missing_values_after_replacement = df.isnull().sum()

# Mostrar el número de valores ausentes después del reemplazo
print("Número de valores ausentes en cada columna después del reemplazo:")
print(missing_values_after_replacement)

# Verificar que no haya valores ausentes en el conjunto de datos
total_missing_values = missing_values_after_replacement.sum()

if total_missing_values == 0:
    print("No hay valores ausentes en el conjunto de datos después del reemplazo.")
else:
    print(f"Aún hay {total_missing_values} valores ausentes en el conjunto de datos.")



Número de valores ausentes en cada columna después del reemplazo:
userid    0
track     0
artist    0
genre     0
city      0
time      0
day       0
dtype: int64
No hay valores ausentes en el conjunto de datos después del reemplazo.


[Volver a Contenidos](#back)

# 2.3 Duplicados <a id='duplicates'></a>
Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos en una secuencia para obtener la cantidad de duplicados explícitos.

In [30]:
# Contar duplicados explícitos

# Calcular el número de duplicados explícitos en la tabla
duplicate_rows = df.duplicated()
number_of_duplicates = duplicate_rows.sum()

# Mostrar el número de duplicados explícitos
print("Número de duplicados explícitos en la tabla:")
print(number_of_duplicates)



Número de duplicados explícitos en la tabla:
3826


Ahora, elimina todos los duplicados. Para ello, llama al método que hace exactamente esto.

In [35]:


# Eliminar todos los duplicados
df = df.drop_duplicates()

# Calcular el número de duplicados explícitos después de eliminarlos
number_of_duplicates_after = df.duplicated().sum()

# Mostrar el número de duplicados explícitos después de eliminarlos
print("Número de duplicados explícitos en la tabla después de eliminarlos:")
print(number_of_duplicates_after)


Número de duplicados explícitos en la tabla después de eliminarlos:
0


Comprobemos ahora si eliminamos con éxito todos los duplicados. Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [36]:
# Comprobar de nuevo si hay duplicados


# Calcular el número de duplicados explícitos después de eliminarlos
number_of_duplicates_after = df.duplicated().sum()

# Mostrar el número de duplicados explícitos después de eliminarlos
print("Número de duplicados explícitos en la tabla después de eliminarlos:")
print(number_of_duplicates_after)

# Comprobar si eliminamos con éxito todos los duplicados
if number_of_duplicates_after == 0:
    print("Todos los duplicados han sido eliminados con éxito.")
else:
    print(f"Aún hay {number_of_duplicates_after} duplicados en la tabla.")



Número de duplicados explícitos en la tabla después de eliminarlos:
0
Todos los duplicados han sido eliminados con éxito.


Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Para hacerlo, primero mostremos una lista de nombres de género únicos, ordenados en orden alfabético. Para ello:
* Extrae la columna `genre` del DataFrame.
* Llama al método que devolverá todos los valores únicos en la columna extraída.


In [37]:
# Inspeccionar los nombres de géneros únicos

# Extraer la columna 'genre'
genre_column = df['genre']

# Obtener los valores únicos en la columna 'genre'
unique_genres = genre_column.unique()

# Ordenar los valores únicos en orden alfabético
unique_genres_sorted = sorted(unique_genres)

# Mostrar los nombres de género únicos ordenados alfabéticamente
print("Nombres de género únicos ordenados alfabéticamente:")
for genre in unique_genres_sorted:
    print(genre)


Nombres de género únicos ordenados alfabéticamente:
acid
acoustic
action
adult
africa
afrikaans
alternative
ambient
americana
animated
anime
arabesk
arabic
arena
argentinetango
art
audiobook
avantgarde
axé
baile
balkan
beats
bigroom
black
bluegrass
blues
bollywood
bossa
brazilian
breakbeat
breaks
broadway
cantautori
cantopop
canzone
caribbean
caucasian
celtic
chamber
children
chill
chinese
choral
christian
christmas
classical
classicmetal
club
colombian
comedy
conjazz
contemporary
country
cuban
dance
dancehall
dancepop
dark
death
deep
deutschrock
deutschspr
dirty
disco
dnb
documentary
downbeat
downtempo
drum
dub
dubstep
eastern
easy
electronic
electropop
emo
entehno
epicmetal
estrada
ethnic
eurofolk
european
experimental
extrememetal
fado
film
fitness
flamenco
folk
folklore
folkmetal
folkrock
folktronica
forró
frankreich
französisch
french
funk
future
gangsta
garage
german
ghazal
gitarre
glitch
gospel
gothic
grime
grunge
gypsy
handsup
hard'n'heavy
hardcore
hardstyle
hardtechno
hip
hip-

Busca en la lista para encontrar duplicados implícitos del género `hiphop`. Estos pueden ser nombres escritos incorrectamente o nombres alternativos para el mismo género.

Verás los siguientes duplicados implícitos:
* `hip`
* `hop`
* `hip-hop`

Para deshacerte de ellos, crea una función llamada `replace_wrong_genres()` con dos parámetros:
* `wrong_genres=`: esta es una lista que contiene todos los valores que necesitas reemplazar.
* `correct_genre=`: este es un string que vas a utilizar como reemplazo.

Como resultado, la función debería corregir los nombres en la columna `'genre'` de la tabla `df`, es decir, remplazar cada valor de la lista `wrong_genres` por el valor en `correct_genre`.

Dentro del cuerpo de la función, utiliza un bucle `'for'` para iterar sobre la lista de géneros incorrectos, extrae la columna `'genre'` y aplica el método `replace` para hacer correcciones.

In [27]:
# Función para reemplazar duplicados implícitos
def replace_wrong_genres(df, wrong_genres, correct_genre):
    for wrong_genre in wrong_genres:
        df['genre'] = df['genre'].replace(wrong_genre, correct_genre)

Ahora, llama a `replace_wrong_genres()` y pásale tales argumentos para que retire los duplicados implícitos (`hip`, `hop` y `hip-hop`) y los reemplace por `hiphop`:

In [39]:
# Eliminar duplicados implícitos



# Llamar a la función para reemplazar los duplicados implícitos
wrong_genres = ['hip', 'hop', 'hip-hop']
correct_genre = 'hiphop'
replace_wrong_genres(df, wrong_genres, correct_genre)

# Comprobar y mostrar la lista de valores únicos de la columna 'genre' una vez más
unique_genres_after_replacement = sorted(df['genre'].unique())

print("Nombres de género únicos ordenados alfabéticamente después de eliminar duplicados implícitos:")
for genre in unique_genres_after_replacement:
    print(genre)



Nombres de género únicos ordenados alfabéticamente después de eliminar duplicados implícitos:
acid
acoustic
action
adult
africa
afrikaans
alternative
ambient
americana
animated
anime
arabesk
arabic
arena
argentinetango
art
audiobook
avantgarde
axé
baile
balkan
beats
bigroom
black
bluegrass
blues
bollywood
bossa
brazilian
breakbeat
breaks
broadway
cantautori
cantopop
canzone
caribbean
caucasian
celtic
chamber
children
chill
chinese
choral
christian
christmas
classical
classicmetal
club
colombian
comedy
conjazz
contemporary
country
cuban
dance
dancehall
dancepop
dark
death
deep
deutschrock
deutschspr
dirty
disco
dnb
documentary
downbeat
downtempo
drum
dub
dubstep
eastern
easy
electronic
electropop
emo
entehno
epicmetal
estrada
ethnic
eurofolk
european
experimental
extrememetal
fado
film
fitness
flamenco
folk
folklore
folkmetal
folkrock
folktronica
forró
frankreich
französisch
french
funk
future
gangsta
garage
german
ghazal
gitarre
glitch
gospel
gothic
grime
grunge
gypsy
handsup
hard'n'he

Asegúrate de que los nombres duplicados han sido eliminados. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [40]:
# Comprobación de duplicados implícitos

# Comprobar y mostrar la lista de valores únicos de la columna 'genre' una vez más
unique_genres_after_replacement = sorted(df['genre'].unique())

print("Nombres de género únicos ordenados alfabéticamente después de eliminar duplicados implícitos:")
for genre in unique_genres_after_replacement:
    print(genre)



Nombres de género únicos ordenados alfabéticamente después de eliminar duplicados implícitos:
acid
acoustic
action
adult
africa
afrikaans
alternative
ambient
americana
animated
anime
arabesk
arabic
arena
argentinetango
art
audiobook
avantgarde
axé
baile
balkan
beats
bigroom
black
bluegrass
blues
bollywood
bossa
brazilian
breakbeat
breaks
broadway
cantautori
cantopop
canzone
caribbean
caucasian
celtic
chamber
children
chill
chinese
choral
christian
christmas
classical
classicmetal
club
colombian
comedy
conjazz
contemporary
country
cuban
dance
dancehall
dancepop
dark
death
deep
deutschrock
deutschspr
dirty
disco
dnb
documentary
downbeat
downtempo
drum
dub
dubstep
eastern
easy
electronic
electropop
emo
entehno
epicmetal
estrada
ethnic
eurofolk
european
experimental
extrememetal
fado
film
fitness
flamenco
folk
folklore
folkmetal
folkrock
folktronica
forró
frankreich
französisch
french
funk
future
gangsta
garage
german
ghazal
gitarre
glitch
gospel
gothic
grime
grunge
gypsy
handsup
hard'n'he

[Volver a Contenidos](#back)

# 2.4 Conclusiones <a id='data_preprocessing_conclusions'></a>

Durante el preprocesamiento de los datos, me di cuenta de que los nombres de las columnas necesitaban ser más consistentes. Empecé por normalizarlos, eliminando espacios, poniendo todos los caracteres en minúsculas y cambiando el nombre de la columna userid a user_id para seguir el estilo snake_case. Luego, me enfoqué en los valores ausentes en las columnas 'track', 'artist' y 'genre', reemplazándolos con 'unknown'.

Después, me ocupé de los duplicados explícitos en el DataFrame utilizando el método drop_duplicates(), lo que ayudó a eliminar filas idénticas y a limpiar los datos.

Al examinar la columna 'genre', noté que había duplicados implícitos, es decir, diferentes formas de escribir el mismo género. Por ejemplo, encontré variaciones como 'hip', 'hop' y 'hip-hop' que deberían ser 'hiphop'. Para resolver esto, creé una función llamada replace_wrong_genres() que reemplazó estos nombres incorrectos por 'hiphop'.

Finalmente, verifiqué que todos los duplicados implícitos habían sido eliminados al mostrar una lista de géneros únicos ordenados alfabéticamente. Todo este proceso me ayudó a limpiar y normalizar los datos, asegurando que estuvieran listos para un análisis más preciso. Ahora, los datos son más consistentes y confiables, lo que me permitirá extraer información valiosa y probar hipótesis con mayor exactitud.

[Volver a Contenidos](#back)

# Etapa 3. Prueba de hipótesis <a id='hypothesis'></a>

# 3.1 Hipótesis: comparar el comportamiento del usuario o la usuaria en las dos ciudades <a id='activity'></a>

La hipótesis afirma que existen diferencias en la forma en que los usuarios y las usuarias de Springfield y Shelbyville consumen música. Para comprobar esto, usa los datos de tres días de la semana: lunes, miércoles y viernes.

* Agrupa a los usuarios y las usuarias por ciudad.
* Compara el número de canciones que cada grupo reprodujo el lunes, el miércoles y el viernes.


Realiza cada cálculo por separado.

El primer paso es evaluar la actividad del usuario en cada ciudad. Recuerda las etapas dividir-aplicar-combinar de las que hablamos anteriormente en la lección. Tu objetivo ahora es agrupar los datos por ciudad, aplicar el método apropiado para contar durante la etapa de aplicación y luego encontrar la cantidad de canciones reproducidas en cada grupo especificando la columna para obtener el recuento.

A continuación se muestra un ejemplo de cómo debería verse el resultado final:
`df.groupby(by='....')['column'].method()`Realiza cada cálculo por separado.

Para evaluar la actividad de los usuarios y las usuarias en cada ciudad, agrupa los datos por ciudad y encuentra la cantidad de canciones reproducidas en cada grupo.



In [43]:
# Contar las canciones reproducidas en cada ciudad


# Filtrar los datos para los días lunes, miércoles y viernes
days_of_interest = ['Monday', 'Wednesday', 'Friday']
filtered_df = df[df['day'].isin(days_of_interest)]

# Contar las canciones reproducidas en cada ciudad para cada uno de los días de interés
for day in days_of_interest:
    count_by_city = filtered_df[filtered_df['day'] == day].groupby('city')['track'].count()
    print(f"Canciones reproducidas el {day}:")
    print(count_by_city)
    print("\n")


Canciones reproducidas el Monday:
city
Shelbyville     5614
Springfield    15740
Name: track, dtype: int64


Canciones reproducidas el Wednesday:
city
Shelbyville     7003
Springfield    11056
Name: track, dtype: int64


Canciones reproducidas el Friday:
city
Shelbyville     5895
Springfield    15945
Name: track, dtype: int64




`Comenta tus observaciones aquí`

Lunes (Monday):
En Springfield, se reprodujeron 15,740 canciones.
En Shelbyville, se reprodujeron 5,614 canciones.
Podemos observar que en Springfield se reproduce casi tres veces más música que en Shelbyville los lunes. Esto sugiere una mayor actividad musical en Springfield al inicio de la semana.

Miércoles (Wednesday):
En Springfield, se reprodujeron 11,056 canciones.
En Shelbyville, se reprodujeron 7,003 canciones.
El miércoles, la diferencia en la cantidad de canciones reproducidas entre las dos ciudades se reduce, pero Springfield sigue teniendo una mayor actividad de reproducción musical en comparación con Shelbyville.

Viernes (Friday):
En Springfield, se reprodujeron 15,945 canciones.
En Shelbyville, se reprodujeron 5,895 canciones.
El viernes, Springfield nuevamente muestra una actividad de reproducción musical significativamente mayor que Shelbyville, similar al lunes.

Conclusiones Iniciales:
En general, Springfield tiene una mayor actividad de reproducción musical que Shelbyville en los días analizados (lunes, miércoles y viernes).
La diferencia en la actividad musical entre las dos ciudades es más pronunciada al inicio y al final de la semana (lunes y viernes).
Aunque Shelbyville muestra una mayor actividad el miércoles en comparación con el lunes, sigue estando por debajo de Springfield en todos los días analizados.
Estas observaciones iniciales sugieren que los usuarios de Springfield tienden a reproducir más música que los de Shelbyville en los días de la semana analizados, lo que puede indicar diferencias en el comportamiento de consumo musical entre las dos ciudades. 

Ahora agrupemos los datos por día de la semana y encontremos el número de canciones reproducidas el lunes, miércoles y viernes. Utiliza el mismo método que antes, pero ahora necesitamos una agrupación diferente.


In [45]:
# Calcular las canciones reproducidas en cada uno de los tres días


# Filtrar los datos para los días lunes, miércoles y viernes
days_of_interest = ['Monday', 'Wednesday', 'Friday']
filtered_df = df[df['day'].isin(days_of_interest)]

# Agrupar los datos por día y ciudad y contar las canciones reproducidas
songs_by_day_and_city = filtered_df.groupby(['day', 'city'])['track'].count().unstack()

# Mostrar el número de canciones reproducidas en cada día para cada ciudad
print("Número de canciones reproducidas en cada día para cada ciudad:")
print(songs_by_day_and_city)



Número de canciones reproducidas en cada día para cada ciudad:
city       Shelbyville  Springfield
day                                
Friday            5895        15945
Monday            5614        15740
Wednesday         7003        11056


`Comenta tus observaciones aquí`
Conclusiones Iniciales: 1) Consistencia en la Actividad Musical: En todos los días analizados (lunes, miércoles y viernes), Springfield muestra una actividad musical consistentemente mayor que Shelbyville. 2) Patrón de Consumo Musical: La actividad musical en Springfield es significativamente alta al inicio y al final de la semana (lunes y viernes), mientras que Shelbyville muestra un aumento de actividad el miércoles, aunque sigue siendo menor en comparación con Springfield y 3) Implicaciones: Estas observaciones sugieren que los usuarios de Springfield tienen un patrón de consumo musical más intenso que los de Shelbyville, lo cual puede ser relevante para estrategias de marketing y análisis de comportamiento del usuario.


Ya sabes cómo contar entradas agrupándolas por ciudad o día. Ahora necesitas escribir una función que pueda contar entradas según ambos criterios simultáneamente.

Crea la función `number_tracks()` para calcular el número de canciones reproducidas en un determinado día **y** ciudad. La función debe aceptar dos parámetros:

- `day`: un día de la semana para filtrar. Por ejemplo, `'Monday'` (lunes).
- `city`: una ciudad para filtrar. Por ejemplo, `'Springfield'`.

Dentro de la función, aplicarás un filtrado consecutivo con indexación lógica.

Primero filtra los datos por día y luego filtra la tabla resultante por ciudad.

Después de filtrar los datos por dos criterios, cuenta el número de valores de la columna 'user_id' en la tabla resultante. Este recuento representa el número de entradas que estás buscando. Guarda el resultado en una nueva variable y devuélvelo desde la función.

In [49]:

# Definir la función number_tracks
def number_tracks(day, city):
    # Filtrar los datos por día
    filtered_by_day = df[df['day'].str.lower() == day.lower()]
    # Filtrar los datos resultantes por ciudad
    filtered_by_city = filtered_by_day[filtered_by_day['city'].str.lower() == city.lower()]
    # Contar el número de valores en la columna 'user_id'
    track_count = filtered_by_city['track'].count()
    return track_count

# Ejemplo de uso de la función
monday_springfield_tracks = number_tracks('Monday', 'Springfield')
wednesday_shelbyville_tracks = number_tracks('Wednesday', 'Shelbyville')
friday_springfield_tracks = number_tracks('Friday', 'Springfield')

print(f"Número de canciones reproducidas en Springfield el lunes: {monday_springfield_tracks}")
print(f"Número de canciones reproducidas en Shelbyville el miércoles: {wednesday_shelbyville_tracks}")
print(f"Número de canciones reproducidas en Springfield el viernes: {friday_springfield_tracks}")


Número de canciones reproducidas en Springfield el lunes: 15740
Número de canciones reproducidas en Shelbyville el miércoles: 7003
Número de canciones reproducidas en Springfield el viernes: 15945


Llama a `number_tracks()` seis veces, cambiando los valores de los parámetros para que recuperes los datos de ambas ciudades para cada uno de los tres días.

In [50]:
# El número de canciones reproducidas en Springfield el lunes


# Definir la función para reemplazar duplicados implícitos
def replace_wrong_genres(df, wrong_genres, correct_genre):
    for wrong_genre in wrong_genres:
        df['genre'] = df['genre'].replace(wrong_genre, correct_genre)

# Llamar a la función para reemplazar los duplicados implícitos
wrong_genres = ['hip', 'hop', 'hip-hop']
correct_genre = 'hiphop'
replace_wrong_genres(df, wrong_genres, correct_genre)

# Definir la función number_tracks
def number_tracks(day, city):
    # Filtrar los datos por día
    filtered_by_day = df[df['day'].str.lower() == day.lower()]
    # Filtrar los datos resultantes por ciudad
    filtered_by_city = filtered_by_day[filtered_by_day['city'].str.lower() == city.lower()]
    # Contar el número de valores en la columna 'track'
    track_count = filtered_by_city['track'].count()
    return track_count

# Llamar a la función para obtener el número de canciones reproducidas en Springfield el lunes
monday_springfield_tracks = number_tracks('Monday', 'Springfield')

print(f"Número de canciones reproducidas en Springfield el lunes: {monday_springfield_tracks}")

Número de canciones reproducidas en Springfield el lunes: 15740


In [5]:
# El número de canciones reproducidas en Shelbyville el lunes
import pandas as pd

# Leer el archivo CSV desde la carpeta /datasets/
file_path = '/datasets/music_project_en.csv'
df = pd.read_csv(file_path)

# Normalizar los nombres de las columnas
df.columns = [col.strip().lower() for col in df.columns]

# Reemplazar valores ausentes en las columnas 'track', 'artist', y 'genre' con 'unknown'
columns_to_replace = ['track', 'artist', 'genre']
for col in columns_to_replace:
    df[col].fillna('unknown', inplace=True)

# Eliminar todos los duplicados explícitos
df = df.drop_duplicates()

# Definir la función para reemplazar duplicados implícitos
def replace_wrong_genres(df, wrong_genres, correct_genre):
    for wrong_genre in wrong_genres:
        df['genre'] = df['genre'].replace(wrong_genre, correct_genre)

# Llamar a la función para reemplazar los duplicados implícitos
wrong_genres = ['hip', 'hop', 'hip-hop']
correct_genre = 'hiphop'
replace_wrong_genres(df, wrong_genres, correct_genre)

# Definir la función number_tracks
def number_tracks(day, city):
    # Filtrar los datos por día
    filtered_by_day = df[df['day'].str.lower() == day.lower()]
    # Filtrar los datos resultantes por ciudad
    filtered_by_city = filtered_by_day[filtered_by_day['city'].str.lower() == city.lower()]
    # Contar el número de valores en la columna 'track'
    track_count = filtered_by_city['track'].count()
    return track_count

# Llamar a la función para obtener el número de canciones reproducidas en Shelbyville el lunes
monday_Shelbyville_tracks = number_tracks('Monday', 'Shelbyville')

print(f"Número de canciones reproducidas en Shelbyville el lunes: {monday_Shelbyville_tracks}")


Número de canciones reproducidas en Shelbyville el lunes: 5614


In [51]:
# El número de canciones reproducidas en Springfield el miércoles


# Definir la función number_tracks
def number_tracks(day, city):
    # Filtrar los datos por día
    filtered_by_day = df[df['day'].str.lower() == day.lower()]
    # Filtrar los datos resultantes por ciudad
    filtered_by_city = filtered_by_day[filtered_by_day['city'].str.lower() == city.lower()]
    # Contar el número de valores en la columna 'track'
    track_count = filtered_by_city['track'].count()
    return track_count

# Llamar a la función para obtener el número de canciones reproducidas en Springfield el miércoles
wednesday_springfield_tracks = number_tracks('Wednesday', 'Springfield')

print(f"Número de canciones reproducidas en Springfield el miércoles: {wednesday_springfield_tracks}")



Número de canciones reproducidas en Springfield el miércoles: 11056


In [52]:
# El número de canciones reproducidas en Shelbyville el miércoles


# Definir la función number_tracks
def number_tracks(day, city):
    # Filtrar los datos por día
    filtered_by_day = df[df['day'].str.lower() == day.lower()]
    # Filtrar los datos resultantes por ciudad
    filtered_by_city = filtered_by_day[filtered_by_day['city'].str.lower() == city.lower()]
    # Contar el número de valores en la columna 'track'
    track_count = filtered_by_city['track'].count()
    return track_count

# Llamar a la función para obtener el número de canciones reproducidas en Shelbyville el miércoles
wednesday_shelbyville_tracks = number_tracks('Wednesday', 'Shelbyville')

print(f"Número de canciones reproducidas en Shelbyville el miércoles: {wednesday_shelbyville_tracks}")



Número de canciones reproducidas en Shelbyville el miércoles: 7003


In [53]:
# El número de canciones reproducidas en Springfield el viernes

# Definir la función number_tracks
def number_tracks(day, city):
    # Filtrar los datos por día
    filtered_by_day = df[df['day'].str.lower() == day.lower()]
    # Filtrar los datos resultantes por ciudad
    filtered_by_city = filtered_by_day[filtered_by_day['city'].str.lower() == city.lower()]
    # Contar el número de valores en la columna 'track'
    track_count = filtered_by_city['track'].count()
    return track_count

# Llamar a la función para obtener el número de canciones reproducidas en Springfield el viernes
friday_springfield_tracks = number_tracks('Friday', 'Springfield')

print(f"Número de canciones reproducidas en Springfield el viernes: {friday_springfield_tracks}")


Número de canciones reproducidas en Springfield el viernes: 15945


In [54]:
# El número de canciones reproducidas en Shelbyville el viernes

# Definir la función number_tracks
def number_tracks(day, city):
    # Filtrar los datos por día
    filtered_by_day = df[df['day'].str.lower() == day.lower()]
    # Filtrar los datos resultantes por ciudad
    filtered_by_city = filtered_by_day[filtered_by_day['city'].str.lower() == city.lower()]
    # Contar el número de valores en la columna 'track'
    track_count = filtered_by_city['track'].count()
    return track_count

# Llamar a la función para obtener el número de canciones reproducidas en Shelbyville el viernes
friday_shelbyville_tracks = number_tracks('Friday', 'Shelbyville')

print(f"Número de canciones reproducidas en Shelbyville el viernes: {friday_shelbyville_tracks}")


Número de canciones reproducidas en Shelbyville el viernes: 5895


**Conclusiones**

`Comenta si la hipótesis es correcta o se debe rechazar. Explica tu razonamiento.`

[Volver a Contenidos](#back)

# Conclusiones <a id='end'></a>

Tras analizar los datos de reproducción de canciones en Springfield y Shelbyville durante los días lunes, miércoles y viernes, llegué a las siguientes conclusiones:

Diferencias Según la Ciudad:

En todos los días analizados, Springfield muestra una mayor actividad de reproducción musical en comparación con Shelbyville.
El lunes, en Springfield se reproducen 15,740 canciones frente a las 5,614 en Shelbyville.
El miércoles, Springfield registra 11,056 reproducciones, mientras que Shelbyville tiene 7,003.
El viernes, Springfield vuelve a tener una actividad notablemente mayor con 15,945 canciones reproducidas, en contraste con las 5,895 en Shelbyville.

Diferencias Según el Día de la Semana:

En Springfield, la actividad de reproducción musical es alta tanto al inicio como al final de la semana. El lunes se reproducen 15,740 canciones y el viernes 15,945, mientras que el miércoles hay una disminución a 11,056 canciones.
En Shelbyville, aunque la actividad es más uniforme a lo largo de la semana, también se observa una disminución en la cantidad de canciones reproducidas el viernes en comparación con el miércoles. El miércoles se reproducen 7,003 canciones, mientras que el viernes el número baja a 5,895.

Confirmación de la Hipótesis:

La hipótesis planteada establece que "la actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad".
Los resultados apoyan esta hipótesis, mostrando diferencias significativas en la actividad de reproducción de música tanto entre las dos ciudades como entre los diferentes días de la semana.

Razonamiento:

La diferencia en el número de canciones reproducidas entre Springfield y Shelbyville cada día de la semana apoya la parte de la hipótesis relacionada con la ciudad.
La variación en la actividad de reproducción entre lunes, miércoles y viernes dentro de cada ciudad respalda la parte de la hipótesis relacionada con el día de la semana.
Por lo tanto, puedo concluir que la hipótesis es correcta. La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad. Los datos analizados muestran claras diferencias tanto entre las ciudades como entre los días de la semana, validando la hipótesis planteada.

### Nota
En proyectos de investigación reales, la prueba de hipótesis estadística es más precisa y cuantitativa. También ten en cuenta que no siempre se pueden sacar conclusiones sobre una ciudad entera a partir de datos de una sola fuente.

Aprenderás más sobre la prueba de hipótesis en el sprint de análisis estadístico de datos.

[Volver a Contenidos](#back)